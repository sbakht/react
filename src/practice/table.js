import React from 'react';

class Table extends React.Component {
  render() {
    const grammar = this.props.grammar;
    return (
      <div>
        <table>
          <thead>
            <tr>Fil</tr>
            <tr>Mudari</tr>
            <tr>Masdr</tr>
            <tr>Faail</tr>
          </thead>
          <tbody>
            <Active grammar={grammar} />
            <Passive grammar={grammar} />
          </tbody>
        </table>
        <Command grammar={grammar} />
      </div>
    );
  }
}

class Active extends React.Component {
  render() {
    const grammar = this.props.grammar.word;
    return (
      <tr id="active">
        <td>
          {grammar.past}
        </td>
        <td>
          {grammar.present}
        </td>
        <td>
          {grammar.masdr}
        </td>
        <td>
          {grammar.faail}
        </td>
      </tr>
    );
  }
}

class Passive extends React.Component {
  render() {
    if (this.props.grammar.hasPassive) {
      return (
        <tr id="passive">
          <td>
            {this.props.grammar.word.pastP}
          </td>
          <td>
            {this.props.grammar.word.presentP}
          </td>
          <td>
            {this.props.grammar.word.masdr}
          </td>
          <td>
            {this.props.grammar.word.mafool}
          </td>
        </tr>
      );
    }
    return <div>NO PASSIVE</div>;
  }
}

class Command extends React.Component {
  render() {
    const grammar = this.props.grammar.word;
    return (
      <table>
        <tr id="command">
          <td>
            {grammar.command}
          </td>
          <td>
            {grammar.forbid}
          </td>
          <td>
            {grammar.tharf}
          </td>
        </tr>
      </table>
    );
  }
}

export default Table;
