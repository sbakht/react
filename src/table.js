import _ from 'underscore';
import Dictionary from './util/dictionary';

var endings = Dictionary.verb.simple.endings;
var d = 'ُ';
var f = 'َ';
var k = 'ِ';
var s = 'ْ';
var sh ='ّ';

function replacer(i) {
  if(i === 0) {
    return '';
  }
  if(i === 1) {
    return d;
  }
  if(i === 2) {
    return f;
  }
  if(i === 3) {
    return k;
  }
  if(i === 4) {
    return s;
  }
  if(i === 5) {
    return sh;
  }
}

class Table {
  constructor(options) {
    this.words = {};
    this.type = options.type;
    if(options.letters) {
      this.build(options.letters, options.type);
    }
  }

  build(str, type) {
    this.buildActive(str, type)
    this.buildPassive(str, type)
  }

  mappings(firstVowels) {
    firstVowels = firstVowels || Dictionary.verb.simple.maps[this.type].beginners;
    var mapping = Dictionary.verb.simple.maps[this.type].enders;

    var completeMapping = mapping.map(m => firstVowels.concat(m));
    return completeMapping.map(m => m.map(replacer));
  }

  buildActive(str, type) {
    var words = [];
    var vowels = this.mappings();

    endings.map(function(ending, i) {
      var word = str + ending;
      var useVowels = vowels[i] || ['','',''];
      words.push(_.flatten(_.zip(word.split(''), useVowels)).join(''));
    });

    this.words.active = words;
  }

  buildPassive(str, type) {
    var words = [];
    var vowels = this.mappings([1,3])

    endings.map(function(ending, i) {
      var word = str + ending;
      var useVowels = vowels[i] || ['','',''];
      words.push(_.flatten(_.zip(word.split(''), useVowels)).join(''));
    });

    this.words.passive = words;
  }
}

export { Table }