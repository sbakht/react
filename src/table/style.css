table {
  direction: rtl;
  width: 100%;
  margin: 0 auto;
}

td {
  font-size: 50px;
}

th {
	text-align: right;
}

form button {
	margin: 0;
}
