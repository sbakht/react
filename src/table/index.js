import React from 'react';
import { Table as TableBuilder } from '../table';
import css from './style.css';


class Table extends React.Component {
	constructor() {
		super();
		this.state = {
			type: "type1",
			roots: "فعل",
			options : { letters : "فعل", group : "maadhi", type : "type1" }
		}

		this.updateRoots = this.updateRoots.bind(this);
		this.updateType = this.updateType.bind(this);
		this.onSubmit = this.onSubmit.bind(this);
	}

	updateRoots(e) {
		this.setState({ roots : e.target.value});
	}
	updateType(e) {
		this.setState({type : e.target.value});
	}
	onSubmit(e) {
		e.preventDefault();
		var options = this.state.options;
		options.type = this.state.type.split('-')[0];
		options.advanced = this.state.type.split('-').length === 2;
		options.letters = this.state.roots;
		this.setState({ options })
	}

	render() {
		var options = this.state.options;
		options.group = "maadhi";
		var table = new TableBuilder(options);
		var maadhi = table.words;

		options.group = "mudari";
		var table = new TableBuilder(options);
		var mudari = table.words;

		var words = [];
		for(var i = 0; i < 15; i++) {
			words.push(
				<tr>
					<td>{maadhi.active[i]}</td>
					<td>{maadhi.passive[i]}</td>
					<td>{mudari.active[i]}</td>
					<td>{mudari.passive[i]}</td>
				</tr>
				)
		}

		return (
			<div>
				<form onSubmit={this.onSubmit}>
					<div>
						<label for="roots">Roots</label>
						<input type="text" value={this.state.roots} onChange={this.updateRoots} />
					</div>

					<div>
						<label for="roots">Type</label>
						<select value={this.state.type} onChange={this.updateType}>
							<optgroup label="Simple">
								<option value="type1">nasara</option>
								<option value="type2">daraba</option>
								<option value="type3">sameya</option>
								<option value="type4">fataha</option>
								<option value="type5">karuma</option>
								<option value="type6">haseba</option>
							</optgroup>
							<optgroup label="Advanced">
								<option value="type1-a">akrama</option>
								<option value="type2-a">sarrafa</option>
								<option value="type3-a">qaatala</option>
								<option value="type4-a">taqabbala</option>
								<option value="type5-a">taqaabala</option>
								<option value="type6-a">ijtanaba</option>
								<option value="type7-a">istansara</option>
								<option value="type8-a">infatara</option>
							</optgroup>
						</select>
					</div>
					<div>
						<button type="submit">Submit</button>
					</div>
				</form>
				<table>
					<thead>
						<tr>
							<th className="grouping" colSpan="2">Maadhi</th>
							<th className="grouping" colSpan="2">Mudari</th>
						</tr>
						<tr>
							<th>Active</th>
							<th>Passive</th>
							<th>Active</th>
							<th>Passive</th>
						</tr>
					</thead>
					<tbody>
						{words}
					</tbody>
				</table>
			</div>
		);
	}
}

export default Table;