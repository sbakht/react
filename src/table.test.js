import React from 'react';
import Util from './util';
import { Table } from './table'
import _ from 'underscore';

function tableTests(options) {
    var table;
    beforeEach(() => {
      table = new Table(options);
    });

    describe("maadhi", () => {
        test('active', () => {
          expect(table.words.maadhi.active).toMatchSnapshot();
        });

        test('passive', () => {
          expect(table.words.maadhi.passive).toMatchSnapshot();
        });
    });

    describe("mudari", () => {
        test('active', () => {
          expect(table.words.mudari.active).toMatchSnapshot();
        });

        test('passive', () => {
          expect(table.words.mudari.passive).toMatchSnapshot();
        });
    });
}

describe("simple", () => {

  describe("nasara", () => {
      var options = { letters : "فعل", type : "type1"}
      tableTests(options);
  });

  describe("daraba", () => {
      var options = { letters : "فعل", type : "type2"}
      tableTests(options);
  });
  describe("sameya", () => {
      var options = { letters : "فعل", type : "type3"}
      tableTests(options);
  });
  describe("fataha", () => {
      var options = { letters : "فعل", type : "type4"}
      tableTests(options);
  });
  describe("karuma", () => {
      var options = { letters : "فعل", type : "type5"}
      tableTests(options);
  });
  describe("haseba", () => {
      var options = { letters : "فعل", type : "type6"}
      tableTests(options);
  });
  
});

describe("advanced", () => {

  describe("akrama", () => {
    var options = { letters : "فعل", advanced: true, type : "type1"}
    tableTests(options);
  });

  describe("sarrafa", () => {
      var options = { letters : "فعل", advanced: true, type : "type2"}
      tableTests(options);
  });

  describe("qaatala", () => {
      var options = { letters : "فعل", advanced: true, type : "type3"}
      tableTests(options);
  });

  describe("taqabbala", () => {
      var options = { letters : "فعل", advanced: true, type : "type4"}
      tableTests(options);
  });

  describe("taqaabala", () => {
      var options = { letters : "فعل", advanced: true, type : "type5"}
      tableTests(options);
  });

  describe("ijtanaba", () => {
      var options = { letters : "فعل", advanced: true, type : "type6"}
      tableTests(options);
  });

  describe("istansara", () => {
      var options = { letters : "فعل", advanced: true, type : "type7"}
      tableTests(options);
  });

  describe("infatara", () => {
      var options = { letters : "فعل", advanced: true, type : "type8"};
      tableTests(options);
  });

});
