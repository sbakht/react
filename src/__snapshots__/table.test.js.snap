// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`builds simple maadhi table 1`] = `
Array [
  "فَعَلَ",
  "فَعَلاَ",
  "فَعَلُوا",
  "فَعَلَتْ",
  "فَعَلَتَا",
  "فَعَلْنَ",
  "فَعَلْتَ",
  "فَعَلْتُمَا",
  "فَعَلْتُمْ",
  "فَعَلْتِ",
  "فَعَلْتُمَا",
  "فَعَلْتُنَّ",
  "فَعَلْتُ",
  "فَعَلْنَا",
]
`;

exports[`builds simple maadhi table 2`] = `
Array [
  "فُعِلَ",
  "فُعِلاَ",
  "فُعِلُوا",
  "فُعِلَتْ",
  "فُعِلَتَا",
  "فُعِلْنَ",
  "فُعِلْتَ",
  "فُعِلْتُمَا",
  "فُعِلْتُمْ",
  "فُعِلْتِ",
  "فُعِلْتُمَا",
  "فُعِلْتُنَّ",
  "فُعِلْتُ",
  "فُعِلْنَا",
]
`;
